kmcd-data.md

```diff
@@ -137,38 +166,40 @@ We model everything with arbitrary precision rationals, but use sort information
                   | Rad  "==Rad" Rad [function]
                   | Rad "=/=Rad" Rad [function]
  // -------------------------------------------
-    rule rad(W1)  <=Rad rad(W2) => W1  <=Rat W2
-    rule rad(W1)   <Rad rad(W2) => W1   <Rat W2
-    rule rad(W1)  >=Rad rad(W2) => W1  >=Rat W2
-    rule rad(W1)   >Rad rad(W2) => W1   >Rat W2
-    rule rad(W1)  ==Rad rad(W2) => W1  ==Rat W2
-    rule rad(W1) =/=Rad rad(W2) => W1 =/=Rat W2
+    rule W1  <=Rad W2 => W1  <=FInt W2
+    rule W1   <Rad W2 => W1   <FInt W2
+    rule W1  >=Rad W2 => W1  >=FInt W2
+    rule W1   >Rad W2 => W1   >FInt W2
+    rule W1  ==Rad W2 => W1  ==FInt W2
+    rule W1 =/=Rad W2 => W1 =/=FInt W2
 ```
 
 ```k
     syntax Rad ::= Wad "*Rate" Ray [function]
  // -----------------------------------------
-    rule wad(R1) *Rate ray(R2) => rad(R1 *Rat R2)
+    rule FInt(W, WAD) *Rate FInt(R, RAY) => FInt(W *Int R, RAD)
 
     syntax Wad ::= Rad "/Rate" Ray [function]
  // -----------------------------------------
-    rule rad(R1) /Rate ray(R2) => wad(R1 /Rat R2)
-
-    syntax Wad ::= rmul ( Wad , Ray ) [function]
- // --------------------------------------------
-    rule rmul(wad(R1), ray(R2)) => wad(R1 *Rat R2)
+    rule FInt(_ , _  ) /Rate ray(0)        => wad(0)
+    rule FInt(R1, RAD) /Rate FInt(R2, RAY) => FInt(R1 /Int R2, WAD) [owise]
 
-    syntax Rad ::= rmul ( Rad , Ray ) [function]
- // --------------------------------------------
-    rule rmul(rad(R1), ray(R2)) => rad(R1 *Rat R2)
+    // syntax Wad ::= rmul ( Wad , Ray ) [function]
+    // syntax Rad ::= rmul ( Rad , Ray ) [function]
+    syntax FInt ::= rmul ( FInt , FInt ) [function]
+ // -----------------------------------------------
+    rule rmul(FInt(W, WAD), FInt(R, RAY)) => FInt((W *Int R) /Int RAY, WAD)
+    rule rmul(FInt(W, RAD), FInt(R, RAY)) => FInt((W *Int R) /Int RAY, RAD)
 
     syntax Ray ::= rdiv ( Ray , Rad ) [function]
```

Comment:
In the codebase, `rdiv` is a thing that takes a `Ray` for its second arg and returns the type of the first arg.

Reply:
Not sure if this is being used anywhere...

Reply:
I just made `rmul`, `rdiv`, and `wdiv` directly go to the underlying `FInt` operators, except in division by zero case.

This means we don't get strong type-checking at uses of these operators (which we didn't get anyway), but we get the correct implementations (where the width of the first operand is always what's returned).